keps/sig-api-machinery/0032-20181026-flexible-inflight-request-limit.md

```diff
@@ -0,0 +1,153 @@
+    * [How to treat namespace-scoped lists?](#how-to-treat-namespace-scoped-lists)
+    * [Should we  make users connect to all apiservers to use their entire allotted qps?](#should-we-make-users-connect-to-all-apiservers)
+
+## Summary <a name="summary"></a>
+This KEP proposes to implement readonly & mutating `capacity` for the apiserver; readonly & mutating `limit` and `request` of requests for different users and groups, considering different physical resource consumption by particular operations (e.g. get single pod vs. get all pods). `limit` limits total inflight request `tokens` from certain user or group and `request` reserves some amount of `tokens` for critical users or groups.
+
+
+## Motivation <a name="motivation"></a>
+
+Buggy service or malicious user may easily run out of '--max-mutating-requests-inflight' or '--max-requests-inflight' request limit by DOSing the apiserver. It will block actually important users (kubelets, controllers/schedulers), that, potentially, leads to impossibility to stop buggy service or block attacker via k8s.
+
+Using only `limit` to prevent blocking important users isn't enough flexible: to guarantee service availability for this users we should set rigid limits, when, sum of all limits not exceed server capacity (e.g. --max-requests-inflight). It will lead to throttling particular user or group even if the server not loaded by other users.
+
+Combination of `limit` and `request` allows to prevent overreservation of resources for critical users on the one hand, and flexible restrict different parts of the system from overspending resources on the other hand.
+
+Using limits in `tokens` instead of limiting number of requests allows to
+account resource consumption more realistic.
+
+
+### Goals <a name="goals"></a>
+
+* Implement per-user and per-group `limit` and `request`
+* Implement request cost estimation
+
+## Proposal <a name="proposal"></a>
+
+Implement weight-based per-user and per-group limits instead of current count-based limits.
+Every request has its own `weight` measured in `tokens`, according to its complexity. Exact algorithm of request cost calculation described below.
+
+Implement options `readonly_capacity`, `mutating_capacity`: Total `weight` of inflight requests that apiserver could carry.
+
+Implement new API resource `InflightLimit` to store group limits and requests:
+```
+apiVersion: extensions/v1beta1
+kind: InflightLimit
+metadata:
+  name: %NAME%
+spec:
+  subjects:
+    kind: %SUBJECT_KIND%
+    name: %SUBJECT_NAME%
+  readonly:
+    limit: %READONLY_TOKENS_LIMIT%
+    request: %READONLY_TOKENS_REQUEST%
+  mutating:
+    limit: %MUTATING_TOKENS_LIMIT%
+    request: %MUTATING_TOKENS_REQUEST%
+status:
+  readonlyInflightTokens: %CURRENT_READONLY_TOKENS_INFLIGHT%
+  mutatingInflightTokens: %CURRENT_MUTATING_TOKENS_INFLIGHT%
+```
+
+Where:
+
+`%NAME%` - The name of `InflightLimit` resource
+
+`%SUBJECT_KIND%` - `User` or `Group`, like in `RoleBinding`
```

Comment:
I've added a comment that is somehow inlined with this one, and apparently this wasn't addressed:
https://github.com/kubernetes/community/pull/2861#discussion_r228866424

I was thinking that it should be uniquely defined - if it's not, then the KEP should contain description of the semantic of that.

Reply:
@liggitt 
I think, it's better to allow only one policy per username. In case of multiple limits defined by several groups, we should pick max limit among user groups, as it described in ["Algorithm to accept/reject readonly request"](https://github.com/kubernetes/community/blame/fa7602b2f9fb44b1f05adcd7dd9f2844e817ade5/keps/sig-api-machinery/0032-20181026-flexible-inflight-request-limit.md#L110) section:

>  Otherwise, among groups of the `$CURRENT_USER`, pick `InflightLimit` with max `readonly.limit` as `$CURRENT_INFLIGHTLIMIT_LIMIT` and `InflightLimit` with max `readonly.request` as `$CURRENT_INFLIGHTLIMIT_REQUEST`.

**motivation:** E.g.: user `admin123` belongs to groups `system:authenticated` and `system:masters`. Obviously, we should use limits & requests defined for `system:masters` for this user.

@wojtek-t 
Sorry if it wasn't clear (or, if i incorrectly understood your concern): i've rewrote KEP to add possibility to set request and limit to given user, not only to group: [Add per-user limits description, clarify algorithm](https://github.com/kubernetes/community/pull/2861/commits/fa7602b2f9fb44b1f05adcd7dd9f2844e817ade5).

In my opinion, defining limits to every given users without any possibility to set limit to whole group will be too hard to maintain for a cluster admin, so it's better to allow to define limits/requests for groups _and_ single users.

Reply:
The approach that SIG API machinery has previously discussed is a two step process:

1. Map (somehow) users / groups / service accounts / namespaces / ??? to buckets
2. Assign quota to buckets.

Anything that is not this general shape is going to be hard to convince @deads2k and me of :)

Reply:
+1 to @lavalamp 
Apparently my comment wasn't clear. For a given request I really would like to see a mapping to a bucket that has quotas. If a request can be mapped to multiple buckets (which my initial comment was supposed to be about), I would really like to see a very detailed description of way to share tokens fairly, as I don't know how this could be done.